src/phoenix/server/api/dataloaders/dataset_example_revisions.py

```diff
@@ -0,0 +1,101 @@
+
+    async def _load_fn(self, keys: List[Key]) -> List[Result]:
+        # sqlalchemy has limited SQLite support for VALUES, so use UNION ALL instead.
+        # For details, see https://github.com/sqlalchemy/sqlalchemy/issues/7228
+        keys_subquery = union_all(
+            *(
+                select(
+                    literal(example_id, Integer).label("example_id"),
+                    literal(version_id, Integer).label("version_id"),
+                )
+                for example_id, version_id in keys
+            )
+        ).subquery()
+        revision_ids = (
+            select(
+                keys_subquery.c.example_id,
+                keys_subquery.c.version_id,
+                func.max(models.DatasetExampleRevision.id).label("revision_id"),
+            )
+            .select_from(keys_subquery)
+            .join(
+                models.DatasetExampleRevision,
+                onclause=keys_subquery.c.example_id
+                == models.DatasetExampleRevision.dataset_example_id,
+            )
+            .where(
+                or_(
+                    keys_subquery.c.version_id.is_(None),
+                    models.DatasetExampleRevision.dataset_version_id <= keys_subquery.c.version_id,
+                )
+            )
+            .group_by(keys_subquery.c.example_id, keys_subquery.c.version_id)
+        ).subquery()
+        query = (
+            select(
+                revision_ids.c.example_id,
+                revision_ids.c.version_id,
+                case(
+                    (
+                        or_(
+                            revision_ids.c.version_id.is_(None),
+                            models.DatasetVersion.id.is_not(None),
+                        ),
+                        True,
+                    ),
+                    else_=False,
+                ).label("is_valid_version"),  # check that non-null versions exist
+                models.DatasetExampleRevision,
+            )
+            .select_from(revision_ids)
+            .join(
+                models.DatasetExampleRevision,
+                onclause=revision_ids.c.revision_id == models.DatasetExampleRevision.id,
+            )
+            .join(
+                models.DatasetVersion,
+                onclause=revision_ids.c.version_id == models.DatasetVersion.id,
+                isouter=True,  # keep rows where the version id is null
+            )
+            .where(models.DatasetExampleRevision.revision_kind != "DELETE")
+        )
+        async with self._db() as session:
+            results = {
+                (example_id, version_id): revision
+                async for (
+                    example_id,
+                    version_id,
+                    is_valid_version,
+                    revision,
+                ) in await session.stream(query)
+                if is_valid_version
+            }
+        if len(results) < len(keys):
+            raise ValueError("Could not find revision.")
```

Comment:
Should return individual exception objects instead of throwing away the whole batch. See https://strawberry.rocks/docs/guides/dataloaders#errors

Reply:
Nice, addressed in f27d7a5fb029993c0da12ea2817a7fee6495d8e1